core/cache/application.go

```diff
@@ -26,20 +30,39 @@ type Application struct {
 
 	details    ApplicationChange
 	configHash string
+	hashCache  *hashCache
 }
 
-func (m *Application) setDetails(details ApplicationChange) {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	m.details = details
+// Config returns the current application config.
+func (a *Application) Config() map[string]interface{} {
+	a.mu.Lock()
+	a.metrics.ModelConfigReads.Inc()
+	a.mu.Unlock()
+	return a.details.Config
+}
 
-	configHash, err := hash(details.Config)
-	if err != nil {
-		logger.Errorf("invariant error - application config should be yaml serializable and hashable, %v", err)
-		configHash = ""
-	}
-	if configHash != m.configHash {
-		m.configHash = configHash
-		// TODO: publish config change...
+// WatchConfig creates a watcher for the application config.
+func (a *Application) WatchConfig(keys ...string) *ConfigWatcher {
+	return newConfigWatcher(keys, a.hashCache, a.hub, a.topic(applicationConfigChange))
+}
+
+func (a *Application) setDetails(details ApplicationChange) {
+	a.mu.Lock()
+
+	a.details = details
+	hashCache, configHash := newHashCache(details.Config, nil, nil)
```

Comment:
This is lifted from the model implementation, but I am not sure about it.

Reply:
i think having some metrics so we know whether our caches are worthwhile is probably a good idea.

Reply:
Done.